[WK2] Flaky ASSERTION FAILED: m_gamepads.isEmpty() on gamepad/gamepad-polling-access.html

marcoscaceres opened the pull request:
#### a53135c21ca451c30ae8fe5dfeb3dc7f05377304
<pre>
[WK2] Flaky ASSERTION FAILED: m_gamepads.isEmpty() on gamepad/gamepad-polling-access.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=245051">https://bugs.webkit.org/show_bug.cgi?id=245051</a>
rdar://99802167

Reviewed by Brady Eidson and Chris Dumez.

The MockGamepadProvider was keeping mock gamepads alive after the test
was done, which caused the ASSERT.

This patch makes the MockGamepadProvider clear its list of mock gamepads
when each test is done.

Also modernized the gamepad-polling-access.html test at bit to use
promises when waiting on the events.

* LayoutTests/gamepad/gamepad-polling-access-expected.txt:
* LayoutTests/gamepad/gamepad-polling-access.html:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
* Source/WebCore/testing/MockGamepadProvider.cpp:
(WebCore::MockGamepadProvider::clearMockGamepads):
* Source/WebCore/testing/MockGamepadProvider.h:

Canonical link: <a href="https://commits.webkit.org/257345@main">https://commits.webkit.org/257345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/546c91f656dc198dabbd93abb5f373121d185087

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98688 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108108 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168360 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102639 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8410 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85267 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91221 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104369 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89930 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33370 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88185 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1820 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1729 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5048 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6674 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42256 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3121 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -149,6 +149,17 @@ void MockGamepadProvider::gamepadInputActivity()
+            disconnectMockGamepad(i);
+    }
+    m_mockGamepadVector.clear();
+    m_mockGamepadVector.resize(0);
```
Not needed since you've just cleared the vector.

cdumez commented on Source/WebCore/testing/MockGamepadProvider.cpp:
```diff
@@ -149,6 +149,17 @@ void MockGamepadProvider::gamepadInputActivity()
 
+void MockGamepadProvider::clearMockGamepads()
+{
+    // disconnect any remaining connected gamepads
```
Comments in WebKit are sentences. They need to start with a capital letter and end with a period.

cdumez reviewed (approved):
LGTM with nits.

marcoscaceres committed:
[WK2] Flaky ASSERTION FAILED: m_gamepads.isEmpty() on gamepad/gamepad-polling-access.html
https://bugs.webkit.org/show_bug.cgi?id=245051
rdar://99802167

Reviewed by Brady Eidson and Chris Dumez.

The MockGamepadProvider was keeping mock gamepads alive after the test
was done, which caused the ASSERT.

This patch makes the MockGamepadProvider clear its list of mock gamepads
when each test is done.

Also modernized the gamepad-polling-access.html test at bit to use
promises when waiting on the events.

* LayoutTests/gamepad/gamepad-polling-access-expected.txt:
* LayoutTests/gamepad/gamepad-polling-access.html:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::Internals):
* Source/WebCore/testing/MockGamepadProvider.cpp:
(WebCore::MockGamepadProvider::clearMockGamepads):
* Source/WebCore/testing/MockGamepadProvider.h:

Canonical link: https://commits.webkit.org/257345@main
